File: tests/src/org.apache-extras.beanshell/bsh/2.0b6/src/test/java/org_apache_extras_beanshell/bsh/ReflectTest.java
/*
 * Copyright and related rights waived via CC0
 *
 * You should have received a copy of the CC0 legalcode along with this
 * work. If not, see <http://creativecommons.org/publicdomain/zero/1.0/>.
 */
package org_apache_extras_beanshell.bsh;

import bsh.Capabilities;
import bsh.Interpreter;

import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.Test;

import static org.assertj.core.api.Assertions.assertThat;

public class ReflectTest {
    @AfterEach
    void resetBeanShellAccessibility() throws Exception {
        Capabilities.setAccessibility(false);
    }

    @Test
    void constructsObjectsWithPublicAndAccessibleConstructorLookup() throws Exception {
        Interpreter interpreter = new Interpreter();

        Capabilities.setAccessibility(false);
        assertThat(interpreter.eval("new String(\"public constructor path\");"))
                .isEqualTo("public constructor path");

        Capabilities.setAccessibility(true);
        assertThat(interpreter.eval("new String(\"declared constructor path\");"))
                .isEqualTo("declared constructor path");
    }

    @Test
    void readsPublicAndPrivateFieldsThroughBeanShellAccess() throws Exception {
        Interpreter interpreter = new Interpreter();
        ReflectiveTarget target = new ReflectiveTarget("visible", "hidden", true);
        interpreter.set("target", target);

        Capabilities.setAccessibility(false);
        assertThat(interpreter.eval("target.publicLabel;"))
                .isEqualTo("visible");

        Capabilities.setAccessibility(true);
        assertThat(interpreter.eval("target.publicLabel;"))
                .isEqualTo("visible");
        assertThat(interpreter.eval("target.privateLabel;"))
                .isEqualTo("hidden");
    }

    @Test
    void invokesPublicAndPrivateMethodsThroughBeanShellAccess() throws Exception {
        Interpreter interpreter = new Interpreter();
        ReflectiveTarget target = new ReflectiveTarget("visible", "hidden", true);
        interpreter.set("target", target);

        Capabilities.setAccessibility(false);
        assertThat(interpreter.eval("target.describe(\"prefix\", 3);"))
                .isEqualTo("prefix:3:visible");

        Capabilities.setAccessibility(true);
        assertThat(interpreter.eval("target.reveal(\"secret\");"))
                .isEqualTo("secret:hidden");
    }

    @Test
    void readsObjectPropertiesThroughGetterFallbacks() throws Exception {
        Interpreter interpreter = new Interpreter();
        ReflectiveTarget target = new ReflectiveTarget("visible", "hidden", true);
        interpreter.set("target", target);

        Capabilities.setAccessibility(false);
        assertThat(interpreter.eval("target.label;"))
                .isEqualTo("visible");
        assertThat(interpreter.eval("target.ready;"))
                .isEqualTo(Boolean.TRUE);
    }

    @Test
    void writesObjectPropertyThroughSetterFallback() throws Exception {
        Interpreter interpreter = new Interpreter();
        ReflectiveTarget target = new ReflectiveTarget("visible", "hidden", false);
        interpreter.set("target", target);

        Capabilities.setAccessibility(false);
        interpreter.eval("target.label = \"updated\";");

        assertThat(target.getLabel()).isEqualTo("updated");
    }

    public static class ReflectiveTarget {
        public String publicLabel;
        private final String privateLabel;
        private final boolean ready;

        public ReflectiveTarget(String publicLabel, String privateLabel, boolean ready) {
            this.publicLabel = publicLabel;
            this.privateLabel = privateLabel;
            this.ready = ready;
        }

        public String describe(String prefix, int count) {
            return prefix + ":" + count + ":" + publicLabel;
        }

        public String getLabel() {
            return publicLabel;
        }

        public void setLabel(String label) {
            this.publicLabel = label;
        }

        public boolean isReady() {
            return ready;
        }

        private String reveal(String prefix) {
            return prefix + ":" + privateLabel;
        }
    }
}

File: tests/src/com.typesafe.akka/akka-actor_2.13/2.8.8/src/test/scala/com_typesafe_akka/akka_actor_2_13/ReflectTest.scala
/*
 * Copyright and related rights waived via CC0
 *
 * You should have received a copy of the CC0 legalcode along with this
 * work. If not, see <http://creativecommons.org/publicdomain/zero/1.0/>.
 */
package com_typesafe_akka.akka_actor_2_13

import java.util.concurrent.CountDownLatch
import java.util.concurrent.TimeUnit
import java.util.concurrent.atomic.AtomicReference

import akka.actor.Actor
import akka.actor.ActorRef
import akka.actor.ActorSystem
import akka.actor.Props
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test

import scala.concurrent.Await
import scala.concurrent.duration.DurationInt

class ReflectTest {

  @Test
  def propsCreatesPublicNoArgumentActorReflectively(): Unit = {
    withActorSystem("reflect-public-no-arg") { system: ActorSystem =>
      val recorder: AtomicReference[String] = new AtomicReference[String]()
      val latch: CountDownLatch = new CountDownLatch(1)

      val actor = system.actorOf(Props(classOf[PublicNoArgActor]), "public-no-arg")
      actor.tell(new ActorMessage("created", recorder, latch), ActorRef.noSender)

      assertThat(latch.await(10, TimeUnit.SECONDS)).isTrue()
      assertThat(recorder.get()).isEqualTo("public-created")
    }
  }

  @Test
  def propsCreatesPrivateNoArgumentActorReflectively(): Unit = {
    withActorSystem("reflect-private-no-arg") { system: ActorSystem =>
      val recorder: AtomicReference[String] = new AtomicReference[String]()
      val latch: CountDownLatch = new CountDownLatch(1)

      val actor = system.actorOf(Props(classOf[PrivateNoArgActor]), "private-no-arg")
      actor.tell(new ActorMessage("created", recorder, latch), ActorRef.noSender)

      assertThat(latch.await(10, TimeUnit.SECONDS)).isTrue()
      assertThat(recorder.get()).isEqualTo("private-created")
    }
  }

  @Test
  def propsCreatesActorWithConstructorArgumentsReflectively(): Unit = {
    withActorSystem("reflect-constructor-args") { system: ActorSystem =>
      val recorder: AtomicReference[String] = new AtomicReference[String]()
      val latch: CountDownLatch = new CountDownLatch(1)

      val props: Props = Props(classOf[ConstructedActor], "constructed", Integer.valueOf(8))
      val actor = system.actorOf(props, "constructed")
      actor.tell(new ActorMessage("created", recorder, latch), ActorRef.noSender)

      assertThat(latch.await(10, TimeUnit.SECONDS)).isTrue()
      assertThat(recorder.get()).isEqualTo("constructed-8-created")
    }
  }

  private def withActorSystem(name: String)(body: ActorSystem => Unit): Unit = {
    val thread: Thread = Thread.currentThread()
    val originalClassLoader: ClassLoader = thread.getContextClassLoader
    thread.setContextClassLoader(null)
    val system: ActorSystem =
      try ActorSystem(name)
      finally thread.setContextClassLoader(originalClassLoader)

    try body(system)
    finally {
      system.terminate()
      Await.result(system.whenTerminated, 10.seconds)
    }
  }
}

final class ActorMessage(
    val value: String,
    val recorder: AtomicReference[String],
    val latch: CountDownLatch)

class PublicNoArgActor extends Actor {
  override def receive: Receive = {
    case message: ActorMessage =>
      message.recorder.set(s"public-${message.value}")
      message.latch.countDown()
  }
}

class PrivateNoArgActor private () extends Actor {
  override def receive: Receive = {
    case message: ActorMessage =>
      message.recorder.set(s"private-${message.value}")
      message.latch.countDown()
  }
}

class ConstructedActor(prefix: String, number: Int) extends Actor {
  override def receive: Receive = {
    case message: ActorMessage =>
      message.recorder.set(s"$prefix-$number-${message.value}")
      message.latch.countDown()
  }
}
